internal/provider/network_connection_data_source.go

```diff
@@ -0,0 +1,66 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	dcsdk "github.com/doublecloud/go-sdk"
+	dcgennet "github.com/doublecloud/go-sdk/gen/network"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+)
+
+func NewNetworkConnectionDataSource() datasource.DataSource {
+	return &NetworkConnectionDataSource{}
+}
+
+// NetworkConnectionDataSource defines the data source implementation.
+type NetworkConnectionDataSource struct {
+	sdk                      *dcsdk.SDK
+	networkConnectionService *dcgennet.NetworkConnectionServiceClient
+}
+
+func (d *NetworkConnectionDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_network_connection"
+}
+
+func (d *NetworkConnectionDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = generateNetworkConnectionDatasourceSchema(resp.Diagnostics)
+}
+
+func (d *NetworkConnectionDataSource) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	sdk, ok := req.ProviderData.(*dcsdk.SDK)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *dcsdk.SDK, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+
+	d.sdk = sdk
+	d.networkConnectionService = d.sdk.Network().NetworkConnection()
+}
+
+func (d *NetworkConnectionDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data *NetworkConnectionModel
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !getNetworkConnection(ctx, d.networkConnectionService, data.ID.ValueString(), data, resp.Diagnostics) {
+		return
```

Comment:
Shall we return an `error` to Diagnostics in case of failed `get` ?

Reply:
`getNetworkConnection` does it for us

Reply:
now this method returns `diag.Diagnostics`